File: 프로그래머스/3/43164. 여행경로/여행경로.js
// function solution(tickets) {
//     const n = tickets.length + 1 // 방문해야하는 총 공항 수 
//   const graph = {}
//   for(let [start, end] of tickets) {
//     if(!graph[start]) {
//       graph[start] = []
//     }
//     if(!graph[end]) {
//       graph[end] = []
//     }
//     // 2번 테스트 케이스(런타임에러)를 통과하기 위해서 end지점도 빈 배열을 생성해줘야합니다.
//     graph[start].push(end)
//     graph[start].sort()
//   }

//   const answer = []
//   const visited = new Set()
//   const dfs = (current, route) => {
//     if(route.length === n) {
//       answer.push([...route])
//       return
//     }
//     for(let i = 0; i < graph[current].length; i++) {
//       const des = graph[current][i]
//       const ticket = `${current}#${des}`
//       if(!visited.has(ticket) || (i >= 1 && ticket === `${current}#${graph[current][i-1]}`)) {
//         // 중복된 티켓팅도 경로를 탐색할 수 있도록 만든 추가 조건문입니다. 1번 테스트 케이스를 위한거 입니다.
//         visited.add(ticket)
//         route.push(des)
//         dfs(des, route)
//         route.pop()
//         visited.delete(ticket)
//       }
//     }
//   }
//   dfs("ICN", ["ICN"])

//   return answer[0]
// }

function solution(tickets){
    const answer = []
    //그래프 그리고
    let graph={}
    for( c of tickets){
        let [start ,end] =c
        graph[start]?graph[start].push(end):graph[start]=[end]
        graph[start].sort() //문자기준 오름차순
    }
    console.log(graph)
    
    const stack = ['ICN']
    while(stack.length>0){
        console.log(stack)
        const now = stack[stack.length-1];
        if (graph[now] && graph[now].length > 0) {//티켓있으면 방문
           stack.push(graph[now].shift())
        } else { //티켓 더이상 없으면   
            console.log('pop',answer)
            answer.unshift(stack.pop());
        }
    }
    return answer
}

File: 프로그래머스/3/43164. 여행경로/여행경로.py
# 4:30~
# ICN공항에서 출발, 방문하는 공항 경로를 []담아서 출력 
# tickects=[[출발,도착],[출,도],...] 3~10000개 주어진 항공권 모두 사용,
# 가능한 경로가 2개 이상이면 알파벳 순서가 앞서는 경로를 return 

# tickets순회하면서 graph ={출발지=[도착지,도착지2], 출발지2:[],...} 만들어준다.
# dfs로 들어가면서
# from collections import deque
# def dfs(graph,cur_node,path):
#     path.append(cur_node)
#     print(cur_node,path)
#     if cur_node not in graph or len(graph[cur_node])==0:
#         return path.copy()
#     else:
#         next_node=graph[cur_node].pop()
#         return dfs(graph,next_node,path)
    

# def solution(tickets):
#     answer = ["ICN"]
#     graph={}
#     for t in tickets:
#         [departure,destination]=t
#         if departure in graph:
#             graph[departure].append(destination)
#         else:
#             graph[departure]=[destination]
#     for g in graph:
#         graph[g]=sorted(graph[g],reverse=True)
        
#     copygraph=graph.copy()
#     answer=dfs(copygraph,'ICN',[])
#     return answer

def solution(tickets):
    routes = {}
    for t in tickets:
        routes[t[0]] = routes.get(t[0], []) + [t[1]]
    for r in routes:
        routes[r].sort(reverse=True)
    stack = ["ICN"]
    path = []
    while len(stack) > 0:
        top = stack[-1]
        if top not in routes or len(routes[top]) == 0:
            path.append(stack.pop())
        else:
            stack.append(routes[top][-1])
            routes[top] = routes[top][:-1]
    return path[::-1]